### PR TITLE
Global logging format changes

### PR DESCRIPTION
**Pull request by peytondmurray**
## Why are these changes needed?

This PR changes how the logging configuration for Ray is set, and changes the format of log messages.

After a discussion with @rkooo567 we've decided to split the logging changes into multiple PRs. This is the first in a series which makes changes to library-level logging code for Ray.

### Changes

* Attempts to consolidate logging configuration by introducing reasonable defaults in `ray/log.py`.
* This new logging configuration is done once in `ray/__init__.py` at the top of the module. Subsequent calls to the configuration are ignored.
* A logger for `ray.rllib` is configured at the WARN level, to address https://github.com/ray-project/ray/pull/31858. With this change,
https://github.com/ray-project/ray/pull/31858 can be reverted, again simplifying and consolidating logging configuration.
* Modified `test_output.py::test_logger_config` to test only the logger config, not launch a ray cluster. The test was failing intermittently, I think due to a race condition between the launch of the cluster and the reading of the subprocess's stdout, and anyway it wasn't necessary to call `ray.init` here to check that logging was configured correctly.
* Modified `python/ray/tune/tests/test_commands.py::test_ls_with_cfg` to test the underlying data, not what gets printed to stdout (which has changed with the new logging system).
* Modified a logging message in `ray.tune.automl.search_policy.AutoMLSearcher.on_trial_complete`, which in certain cases emits a logging message which tries to format a `NoneType` into a `%f` during log message formatting. This was a previously-undetected bug which showed up because the default log level is now `INFO`. This fixes a test that was failing in `test_automl_searcher.py::AutoMLSearcherTest`.

## Related issue number

Partially addresses #30005.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by peytondmurray** [comment]
## Failing tests

Currently there is an issue that causes failures because `pandas` is not installed on the CI runners. This affects every minimal install test suite. 

Default install is also failing, but there's a fix for this available here: #32559.

**Comment by rkooo567** [comment]
Please add me to the assignee section!

**Comment by peytondmurray** [comment]
:point_up: Sorry about that! I saw your name in the assignee section, I should have added you to the reviewers as well!

**Comment by peytondmurray** [comment]
Rebased onto `master` to pull in the new test fixes for rllib.

File: doc/source/ray-observability/ray-logging.rst
Comment (rkooo567) [inline doc/source/ray-observability/ray-logging.rst:9]:
```suggestion
When ``import ray`` is executed, Ray's logger is initialized, generating a sensible configuration given in ``python/ray/log.py``. The default logging level is ``logging.INFO``.
```

File: python/ray/log.py
Comment (rkooo567) [inline python/ray/log.py:1]: Don't create log.py and can you just merge it into ray_logging.py? It is confusing we have 2 files like this

File: python/ray/_private/test_utils.py
Comment (rkooo567) [inline python/ray/_private/test_utils.py:442]: Instead of this, can you just check the env var from the logger or sth? 

We export `BUILDKITE="true"` to the CI tests

File: python/ray/log.py
Comment (rkooo567) [inline python/ray/log.py:55]: whens' the case it is false?

File: python/ray/log.py
Comment (rkooo567) [inline python/ray/log.py:118]: hmm why not just using global variable ?

Comment (rkooo567) [reply python/ray/log.py:118]: logger_initialized = False

def f():
    global logger_initialized
    if logger_initialized:
        return

File: python/ray/log.py
Comment (rkooo567) [inline python/ray/log.py:90]: > This will need to be removed when
        # a proper logging system is set up in a way that lets workers pass
        # LogRecords to the driver.

Can we remove this comment? Still not sure if we will ever do this

File: python/ray/log.py
Comment (rkooo567) [inline python/ray/log.py:187]: how does this work? what's going to happen if user calls the same thing with version 1?

**Review by rkooo567 — COMMENTED** [review]
Can you add unit tests? We should make sure

1. The context is properly printed within each library
2. make sure the logging_level and logging_format works after this PR

I will also play with this log once I come back from OOO (next Mon)

File: python/ray/_private/test_utils.py
Comment (peytondmurray) [reply python/ray/_private/test_utils.py:442]: Done.

File: python/ray/log.py
Comment (peytondmurray) [reply python/ray/log.py:118]: Well, I was thinking that a decorator would be more reusable, and because of concerns about polluting the namespace with global stuff. It also encapsulated the reconfiguration debug message nicely. I'm happy to change this though if there's already an established precedent.

File: python/ray/log.py
Comment (peytondmurray) [reply python/ray/log.py:90]: :+1:

File: python/ray/log.py
Comment (peytondmurray) [reply python/ray/log.py:187]: See https://docs.python.org/3/library/logging.config.html#dictionary-schema-details - python's logging system requires us to pass a `version` key, and its value must be `1` :shrug:.

File: doc/source/ray-observability/ray-logging.rst
Comment (rkooo567) [inline doc/source/ray-observability/ray-logging.rst:25]: Can you add examples for all libraries (data, tune, rllib, air, train, workflow)?

File: python/ray/log.py
Comment (rkooo567) [reply python/ray/log.py:51]: Can you add a comment to the code?

File: python/ray/log.py
Comment (rkooo567) [inline python/ray/log.py:60]: remove it? Seems like it is unused?

File: python/ray/log.py
Comment (rkooo567) [inline python/ray/log.py:66]: Hmm I don't see this is used anywhere?

File: python/ray/log.py
Comment (rkooo567) [inline python/ray/log.py:83]: Maybe we should just use stderr which is the default of regular stream handler?

File: python/ray/log.py
Comment (rkooo567) [inline python/ray/log.py:91]: The docstring seems wrong

File: python/ray/log.py
Comment (rkooo567) [inline python/ray/log.py:114]: Maybe we need to lock this?

File: python/ray/log.py
Comment (rkooo567) [inline python/ray/log.py:127]: do we need this line?

File: python/ray/log.py
Comment (rkooo567) [inline python/ray/log.py:1]: Hmm can you at least move this to _private repo? Or is it also difficult?

**Review by rkooo567 — COMMENTED** [review]
A couple last comments. I will also test this on my local machine tmrw

File: python/ray/log.py
Comment (peytondmurray) [reply python/ray/log.py:83]: Currently we use the logger to provide useful runtime information to the users rather than error reporting, so I'm inclined to keep it `stdout`. What do you think? I'm happy to switch over if you'd prefer.

File: python/ray/log.py
Comment (peytondmurray) [reply python/ray/log.py]: This has been removed as per the request below.

File: doc/source/ray-observability/ray-logging.rst
Comment (peytondmurray) [reply doc/source/ray-observability/ray-logging.rst]: Yep, I added a section showing how to grab handles for each of the ray loggers, and how to set the logging level and add a handler for two different ray subpackages.

File: python/ray/log.py
Comment (peytondmurray) [reply python/ray/log.py]: Yes, this is left over from splitting this PR into two. Thanks for catching this!

File: python/ray/log.py
Comment (peytondmurray) [reply python/ray/log.py]: Again this was left over from splitting the PR. Removed.

File: python/ray/log.py
Comment (peytondmurray) [reply python/ray/log.py]: Removed.

File: python/ray/log.py
Comment (peytondmurray) [reply python/ray/log.py]: Possibly? It's only called once, currently, at `ray` module initialization time. The previous `call_once` decorator effectively implemented a lock, but I can add one here too.

File: python/ray/log.py
Comment (peytondmurray) [reply python/ray/log.py]: No, I think this is okay to remove.

File: python/ray/log.py
Comment (peytondmurray) [reply python/ray/log.py]: Nope, there's no problem with doing this - as long as there are no calls to `import ray` inside the subpackage at initialization time, which there aren't. I've moved the file there now.

File: python/ray/log.py
Comment (peytondmurray) [reply python/ray/log.py]: Switched to `stderr`.

**Comment by peytondmurray** [comment]
After a discussion with @rkooo567, I've modified `ray.init()` to strip the handlers off of the `ray` logger if `configure_logging == False`, which should match the current implementation.

File: python/ray/log.py
Comment (rkooo567) [inline python/ray/log.py:2]: Move this file to _private?

File: python/ray/log.py
Comment (rkooo567) [reply python/ray/log.py:60]: Hmm this still seems to be not removed yet? Is it expected?

File: python/ray/log.py
Comment (rkooo567) [reply python/ray/log.py:51]: seems like it is not removed?

File: python/ray/log.py
Comment (rkooo567) [reply python/ray/log.py:66]: not removed?

File: python/ray/log.py
Comment (rkooo567) [reply python/ray/log.py:91]: also seems to be still here?

File: python/ray/log.py
Comment (rkooo567) [reply python/ray/log.py:114]: yep can you add a lock just in case? should be a very simple to add one

File: python/ray/log.py
Comment (rkooo567) [reply python/ray/log.py:127]: still not removed

**Review by rkooo567 — COMMENTED** [review]
LInt failure. Also, I see some comments were not addressed + agent changes are not here? Is it expected? Maybe something has been screwed by rebase?

**Comment by peytondmurray** [comment]
:point_up: Yep, this got screwed up by a rebase - there was another merge conflict, and I must have messed this up. Will fix this now.

File: python/ray/tests/test_runtime_env_working_dir_2.py
Comment (rkooo567) [inline python/ray/tests/test_runtime_env_working_dir_2.py:150]: hmm maybe we shouldn't delete this line?

**Review by rkooo567 — APPROVED** [review]
Before merging, we need approval on code change from rllib, tune, and serve. 

cc @sven1977 for the rllib change (especially the logging level related since it was regressed last time)

cc @krfricke for the tune test change

cc @sihanwang41 for the serve test changes.

File: doc/source/ray-observability/ray-logging.rst
Comment (sven1977) [inline doc/source/ray-observability/ray-logging.rst:9]: Shouldn't the default logging level be "WARNING"? "INFO" is normally quite chatty (at least for RLlib).

Comment (peytondmurray) [reply doc/source/ray-observability/ray-logging.rst:9]: [The default logging level is `INFO`](https://github.com/peytondmurray/ray/blob/76c47941459f68e074e0a01e0d1c34e0d8c5d8ec/python/ray/_private/log.py#L111) - except for RLlib, [which is `WARNING`](https://github.com/peytondmurray/ray/blob/76c47941459f68e074e0a01e0d1c34e0d8c5d8ec/python/ray/_private/log.py#L117) because of the issue raised in [this PR](https://github.com/ray-project/ray/pull/31858). Maybe I should add a note about this to the documentation?

We currently have a test to ensure that the logging level of ray and each subpackage is not inadvertently modified, so that this doesn't happen in the future either.

**Comment by ollie-iterators** [comment]
Even if it didn't show up in this PR, it would be nice to have the lint tests pass for the PR.

**Comment by rkooo567** [comment]
cc @sven1977 @krfricke @sihanwang41 can you guys review the PR? We hope to merge the PR asap.

**Review by krfricke — APPROVED** [review]
Tune changes LGTM

**Review by sven1977 — APPROVED** [review]
Looks good to me now.
Thanks for all the hard work @peytondmurray ! :)

**Comment by rkooo567** [comment]
@peytondmurray we should respect the logger configured from the library if it is not done that way.

**Comment by peytondmurray** [comment]
Ahh, this was added 2 hours ago, so I didn't see this until now.

> Does it mean lib level can't configure their logging config?

No, all it means is that you can't call `ray._private.log.generate_logging_config` more than once. Any change you make on your own (i.e. anything done outside that function) is respected.

That said, my initial feeling is to move that serve logging configuration stuff to `python/ray/_private/log.py`. Part of the goal of the original logging RFC was to unify logging configuration because it was spread across disparate parts of the code base, making it very difficult to track, modify, and debug. I'd make the case to consolidate logging configuration here, and advise against letting logging configuration leak to different parts of the code base where it gets lost. [We're already doing this with `ray.rllib`](https://github.com/peytondmurray/ray/blob/76c47941459f68e074e0a01e0d1c34e0d8c5d8ec/python/ray/_private/log.py#L116), so we're setting a precedent here. On top of that, if a special handler with emit-time context information is needed, we can do that with the same mechanisms we are already using when we handle plain logging configuration from workers.

**Comment by rkooo567** [comment]
cc @sihanwang41 since your PR has been merged, let's see if this PR will break anything? @peytondmurray can you merge the latest master last time? I think once you can verify the serve logger works, we can get approval and merge it. @sihanwang41 what test should he run to verify this?

**Comment by ollie-iterators** [comment]
If things fail by rebasing to master, it will definitely not be because of this PR. There are major issues with ray happening right now.

**Comment by peytondmurray** [comment]
Rebased and pushed, there were no conflicts. The last version I pushed was passing all tests*, so if this is failing I'm hoping it's not my fault.

*All non-flaky or broken tests, anyhow

**Comment by rkooo567** [comment]
since master is crazy now, let's wait a bit. Also I'd like to merge this PR after 2.4 branch cut...

File: doc/source/ray-observability/ray-logging.rst
Comment (angelinalg) [inline doc/source/ray-observability/ray-logging.rst:11]:
```suggestion
All ray loggers are automatically configured in ``ray._private.ray_logging``. To change the Ray library logging configuration:
```

File: doc/source/ray-observability/ray-logging.rst
Comment (angelinalg) [inline doc/source/ray-observability/ray-logging.rst:20]:
```suggestion
Similarly, to modify the logging configuration for any Ray subcomponent, specify the appropriate logger name:
```

File: doc/source/ray-observability/ray-logging.rst
Comment (angelinalg) [inline doc/source/ray-observability/ray-logging.rst:136]:
```suggestion
Customizing Worker Loggers
```

File: doc/source/ray-observability/ray-logging.rst
Comment (angelinalg) [inline doc/source/ray-observability/ray-logging.rst:208]:
```suggestion
    The default temp directory is ``/tmp/ray`` (for Linux and MacOS). To change the temp directory, specify it when you call ``ray start`` or ``ray.init()``.
```

File: doc/source/ray-observability/ray-logging.rst
Comment (angelinalg) [inline doc/source/ray-observability/ray-logging.rst:233]:
```suggestion
- ``worker-[worker_id]-[job_id]-[pid].[out|err]``: Python/Java part of Ray drivers and workers. All stdout and stderr from tasks or actors are streamed here. Note that job_id is an id of the driver.
```

**Review by angelinalg — APPROVED** [review]

**Comment by peytondmurray** [comment]
I rebased onto `master` again, resolving merge conflicts. Thanks again @angelinalg. @xwjiang2010 I also incorporated your latest changes to the logging system, switching to use `logging._StderrHandler` from `logging.StreamHandler`.

**Pull request merged**
